rubix-tests/src/test/java/com/qubole/rubix/tests/TestNonLocalReadRequestChain.java

```diff
@@ -0,0 +1,167 @@
+import com.qubole.rubix.bookkeeper.LocalDataTransferServer;
+import com.qubole.rubix.core.DataGen;
+import com.qubole.rubix.core.NonLocalReadRequestChain;
+import com.qubole.rubix.core.ReadRequest;
+import com.qubole.rubix.core.TestCachingFileSystem;
+import com.qubole.rubix.spi.CacheConfig;
+import com.qubole.rubix.spi.ClusterType;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.log4j.BasicConfigurator;
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+
+import static org.testng.AssertJUnit.assertTrue;
+
+public class TestNonLocalReadRequestChain
+{
+    int blockSize = 100;
+    String backendFileName = "/tmp/backendFile";
+    Path backendPath = new Path("testfile:/" + backendFileName);
+    File backendFile = new File(backendFileName);
+    final Configuration conf = new Configuration();
+    Thread localDataTransferServer;
+
+    NonLocalReadRequestChain nonLocalReadRequestChain;
+    private static final Log log = LogFactory.getLog(TestNonLocalReadRequestChain.class);
+
+    @BeforeMethod
+    public void setup()
+            throws Exception
+    {
+        conf.setBoolean(CacheConfig.DATA_CACHE_STRICT_MODE, true);
+        conf.setInt(CacheConfig.dataCacheBookkeeperPortConf, 3456);
+        conf.setInt(CacheConfig.localServerPortConf, 2222);
+        conf.setInt(CacheConfig.blockSizeConf, blockSize);
+        localDataTransferServer = new Thread()
+        {
+            public void run()
+            {
+                LocalDataTransferServer.startServer(conf);
+            }
+        };
+        Thread thread = new Thread()
+        {
+            public void run()
+            {
+                BookKeeperServer.startServer(conf);
+            }
+        };
+        thread.start();
+
+        while (!BookKeeperServer.isServerUp()) {
+            Thread.sleep(200);
+            log.info("Waiting for BookKeeper Server to come up");
+        }
+
+        // Populate File
+        DataGen.populateFile(backendFileName);
+
+        //set class for filepath beginning with testfile
+        conf.setClass("fs.testfile.impl", TestCachingFileSystem.class, FileSystem.class);
+        TestCachingFileSystem fs = new TestCachingFileSystem();
+        fs.initialize(null, conf);
+        nonLocalReadRequestChain = new NonLocalReadRequestChain("localhost", backendFile.length(), backendFile.lastModified(), conf, fs, backendPath.toString(), ClusterType.TEST_CLUSTER_MANAGER.ordinal());
+    }
+
+    @Test
+    private void testRemoteRead()
+            throws Exception
+    {
+        localDataTransferServer.start();
```

Comment:
Do one more change to ensure that we are reading from localDataTransferServer and not direct reads here.
Add to NonLocalRRC a check that if STRICT_MODE is set, then it should fail rather than falling back to direct mode